python/gink/impl/database.py

```diff
@@ -268,31 +277,68 @@ def run(self,
             until = self.resolve_timestamp(until)
         context_manager = console or nullcontext()
         with context_manager:
-            while (until is None or generate_timestamp() < until):
-                # eventually will want to support epoll on platforms where its supported
-                readers: List[Union[Listener, Connection, SelectableConsole, AbstractStore]] = []
-                if self._store.is_selectable():
+            # eventually will want to support epoll on platforms where its supported
+            readers: List[Union[Listener, Connection, SelectableConsole, AbstractStore, Socket]] = []
+            if self._store.is_selectable():
                     readers.append(self._store)
-                for listener in self._listeners:
-                    readers.append(listener)
-                for connection in list(self._connections):
-                    if connection.is_closed():
-                        self._connections.remove(connection)
-                    else:
-                        readers.append(connection)
-                if isinstance(console, SelectableConsole):
-                    readers.append(console)
+            for listener in self._listeners:
+                readers.append(listener)
+            for connection in list(self._connections):
+                if connection.is_closed():
+                    self._connections.remove(connection)
+                else:
+                    readers.append(connection)
+            if self._wsgi_server:
+                readers.append(self._wsgi_server.listen_socket)
+            if isinstance(console, SelectableConsole):
+                readers.append(console)
+            while (until is None or generate_timestamp() < until):
                 if console:
                     console.refresh()
                 ready_readers, _, _ = select(readers, [], [], 0.1)
                 for ready_reader in ready_readers:
                     if isinstance(ready_reader, Connection):
                         for data in ready_reader.receive():
                             self._receive_data(data, ready_reader)
+                        if ready_reader.is_closed():
+                            self._connections.remove(ready_reader)
+                            readers.remove(ready_reader)
+                    elif self._wsgi_server and ready_reader == self._wsgi_server.listen_socket:
+                        try:
+                            conn, client_address = self._wsgi_server.listen_socket.accept()
+                        except BlockingIOError as e:
+                            code, msg = e.args
+                            if code == EINTR:
+                                continue
+                            else:
+                                raise e
+                        readers.append(conn)
+                    elif self._wsgi_server and isinstance(ready_reader, Socket):
```

Comment:
I would prefer that we use a wrapper class for WSGI connections, rather than just adding the raw socket to the readers.  Also it would be best to encapsulate a lot of the processing you do here into a class.